### DISCUSSION TITLE
SUSE contributions and future steps

### DISCUSSION THREAD

**Discussion by ggbecker** [ComplianceAsCode/content#6669] [General]
@brett060102 @guangyee

Hello, I would like to start a discussion topic to talk about contributions that recently happened to `sle12` product.

First of all, the contributions were great. They are really valuable for the project and I'm sure it will benefit the community a lot. New content is always appreciated. So far, most of the content is enabled only to Suse12 but they will eventually be enabled for other products as well.

But there were some obstacles that made the pull request review more difficult. For example: lack of tests, pull request granularity, the duplicity of changes among pull requests, etc and I would like to continue discussing how we can improve in the future to avoid them so new contributions can go through more easily.

I'll try to summarize the problems and their solutions or at least what the project's community sees as best practice:

1. Pull requests didn't have any tests.
    - For content related tests, we use the [SSG Test Suite](https://complianceascode.readthedocs.io/en/latest/tests/README.html). It can run scans, remediations, and everything we need to tests OVAL checks and remediations. You can use virtual machines and containers as backends. These tests can help us preventing regressions and can also be used to check the correctness of checks and remediations.
2. Bulk pull requests
    - It's harder to review pull requests that contain many things at once. So more granular pull requests are expected in this case. Even better if they can come at a slow pace, so we have time to react properly. If that's not possible, then we can try to think of a better solution for both sides.
3. Duplicated changes among different pull requests
    - Few pull requests had duplicated changes across different pull requests. This pretty much aligns with the problem above. I guess it was because you had to manually craft the pull requests out of one single commit.
4. Content in some cases was not generic enough
    - This happens because the content was targetting mainly DISA STIG so many of the rules were aligned exclusively with STIG by having hardcoded coded variables, etc. The problems were fixed when requested, but I think it's a good opportunity to better understand the project principles and align better in the future. The main idea here is to create customizable content by using XCCDF variables, so other products/profiles can make use of this shared content. I think you guys already know this, but it's good to reinforce.
5. (update) Force push
    - Although it's tempting to use force push all the time, it can make the reviewer's life much more difficult when he's trying to identify what has changed after he requested changes to the pull request. So in this case, after the initial pull request is created, every new change requested should be included in a separate commit.

If you have any questions or doubts, please do not hesitate and ask them. I'll be more than happy to help.

**Comment by brett060102** [comment]
@ggbecker I have added @abergmann who is also part of our team so that we are on the same page.

Thank you very much for this and for your patience and support on getting these in. We had a steep learning curve on this. I am sorry about the duplicates in the last set of reviews. Our local development branch was about 2 months a head of what we had managed to get pushed upstream and we wanted to get upstream fully in sync with what we had locally and what we had locally in sync with upstream master. The initial try was https://github.com/ComplianceAsCode/content/pull/6629  which was really huge and represented about 30 internal PRs. I could not break it up on a change by change basis because underlying code had changed as we tracked changes in upstream. So, while not ideal, the method I used to break up 6229 was to quash merge the changes in it into another branch and then restore files that I did not want to include in the PR. Which meant storing rough 75% of the files changed by the squash. The process presented a huge opportunity for error and is not something I want to or plan to do again and is really just the result of much we needed to get done in a short period of time.

The process for SL15, will be much more orderly:

1: Final SLE-12 cleanup https://github.com/ComplianceAsCode/content/pull/6665/ Thank you for that review.

2: Initial SL15 support structure, https://github.com/ComplianceAsCode/content/pull/6666

3: PR to add SLE-15 into existing rules that reference SLE-12, but no new content, just adding SLE-15 STIGS, CCE ID's, updating prodtype, and in few places updating checks like {{% if product == "sle12" %}} to {% if product in ['sle12', 'sle15'] %}}.
The PR will change about 190 files. But that seems in line with what was done for RHEL-8. 
There are a few places where we currently have sle12.yml  and sl12.xml files where I needed to create sle15 versions that are just a copy of the sle12 contents. Is there a better way to do that? Create a SUSE specific rule that would get pulled on for both SLE12 and SLE15? The only other use of OS specific files I have seen are for JRE, so not a lot of prior art here. Another option is to merge that into the shared. files. Which I would be OK with, but  would like to cover as separate effort from enabling existing SLE-12 rules for SLE-15.

4: SLE-15 specific content will be done as separate PR's. So normally one rule to a PR. The only exception would be if a STIG spawned multiple rules, which I hope won't be the case.

I still need to digest everything in the testing document, but we will provide tests for all new SLE-15 content and once the SLE-15 is complete go back and add tests in for the new content we created for SLE-12.

I will stop the forced pushes. We do that internally just to keep the review history manageable and make make sure that the SHA1 for changed was complete and not missing some change that was done later. 

Again, thank you for all of your time and guidance on this.

**Reply by ggbecker** [reply]
Thank you very much for the answer.

> 3: PR to add SLE-15 into existing rules that reference SLE-12, but no new content, just adding SLE-15 STIGS, CCE ID's, updating prodtype, and in few places updating checks like {{% if product == "sle12" %}} to {% if product in ['sle12', 'sle15'] %}}.
> The PR will change about 190 files. But that seems in line with what was done for RHEL-8.
> There are a few places where we currently have sle12.yml and sl12.xml files where I needed to create sle15 versions that are just a copy of the sle12 contents. Is there a better way to do that? Create a SUSE specific rule that would get pulled on for both SLE12 and SLE15? The only other use of OS specific files I have seen are for JRE, so not a lot of prior art here. Another option is to merge that into the shared. files. Which I would be OK with, but would like to cover as separate effort from enabling existing SLE-12 rules for SLE-15.

Unfortunately, there is no easy way how to achieve this multiple OS specific check at this moment. You can create a new file which would be a duplicate (not ideal) or embed the content into the shared file but use Jinja macros to keep the code separate. Ideally the best would be to merge everything together and have the jinja macros only when the content actually differs.

**Comment by guangyee** [comment]
@ggbecker Thanks for information! We really appreciate the guidance. Regarding SSG Test Suite, will we be required to expose a SUSE-based VM or container upstream? If so, how does upstream handle product licensing and activation? So far we have been running our tests locally using a SLES12 vagrant box. And license activation is part of provisioning, with an internal developer license code which unfortunately cannot be shared outside of the organization.

While reading through the documentation, I came across this disclaimer. 

"SSG Test Suite currently does not provide automated provisioning of VMs or containers."

Does this mean running the tests will be on-demand by the platform-specific developers? And license activation will be done out-of-band? If so, how does upstream verify the test code? Or is that process also expected to be out-of-band as well?

**Reply by ggbecker** [reply]
> @ggbecker Thanks for information! We really appreciate the guidance. Regarding SSG Test Suite, will we be required to expose a SUSE-based VM or container upstream? If so, how does upstream handle product licensing and activation? So far we have been running our tests locally using a SLES12 vagrant box. And license activation is part of provisioning, with an internal developer license code which unfortunately cannot be shared outside of the organization.

We won't require any exposure of VMs or container upstream, at least for now. So, you don't need to worry about licensing.

> 
> While reading through the documentation, I came across this disclaimer.
> 
> "SSG Test Suite currently does not provide automated provisioning of VMs or containers."
> 
> Does this mean running the tests will be on-demand by the platform-specific developers? And license activation will be done out-of-band? If so, how does upstream verify the test code? Or is that process also expected to be out-of-band as well?

At this moment, we expect everything to be out-of-band. So tests are run on demand by platform-specific developers, as you mentioned. We plan in the future to have some kind of automation in this regard, for example: using Github actions, spawn a container, and run the tests on this container, mainly for pull requests. We are still behind this scenario, so the motivation here is to leverage the test suite we have and create tests we are familiar with so they can be used by you (they will help you with testing and maintaining), and can be used by the community. For example, when your new content is enabled on a different platform, we can easily verify that it does what it intends to do by running the tests. Of course, there will be some exceptions and this will be evaluated on a case-by-case basis. 

The test suite itself might not be 100% applicable to your environment as of now, we may have to adapt/add a few things before you can start using it. We expect that this conversation will keep going and that we can find the best solution for everybody.

**Reply by matejak** [reply]
Thank you for your interest in CI! That one is multi-layered:

1. PR gating - currently, pull requests are gated by building the content for certain products on Jenkins nodes. Currently we have Fedora and RHEL nodes: https://jenkins.complianceascode.io/computer Those nodes now run test for SLE12 and SLE15 products, and those tests make sure that content built on those nodes passes sanity checks - i.e. that the build completes, and the result passes the validation.
In the case that you would be so kind to host a VPS with a SLE system set up as a Jenkins node, we would gladly integrate it in our CI pipeline.
2. Rule tests - as you have probably noticed, rules can be tested individually - that's what the `tests` folder is there for. Rule tests are somehow slow, and there are many rules, so we don't do that as part of the PR gating. We have a bot that [advises users](https://github.com/ComplianceAsCode/content/pull/6453#issuecomment-738127318) to locally tests a PR using the upstream test suite - the bot created by our great engineer @mildas tries to estimate which tests are relevant in the context of a PR, and we hope that we will be able to run them as part of gating later.
The test suite may be somehow difficult to approach, and you can look forward a post going into greater detail on our [upstream blog](https://complianceascode.github.io/) in a couple of days. In a nutshell, the test suite is able to use OS container images or libvirt domains to provision the system using Bash scripts, and to execute the scan, collect the results, and propagate them to the user. So the idea is that the developer sets up the base system once, and then runs the test suite locally as a development support, or runs it overnight for all rules to makes sure that the datastream is in a good shape.

**Reply by brett060102** [reply]
@ggbecker trying to build the stig_overlay and hitting an error:
vagrant@compliance-as-code:~/ComplianceAsCodeContent> PYTHONPATH=`./.pyenv.sh` ./utils/create-stig-overlay.py --disa-xccdf=/tmp/U_SLES_12_STIG_V2R1_Manual-xccdf.xml --ssg-xccdf=build/ssg-sle12-xccdf.xml -o sle12/overlays/stig_overlay.xml
Traceback (most recent call last):
  File "./utils/create-stig-overlay.py", line 171, in <module>
    main()
  File "./utils/create-stig-overlay.py", line 167, in main
    new_stig_overlay(disa_xccdftree, ssg_xccdftree, args.output_file)
  File "./utils/create-stig-overlay.py", line 121, in new_stig_overlay
    dom = xml.dom.minidom.parseString(ET.tostring(new_stig_overlay, encoding="UTF-8", xml_declaration=True))
TypeError: tostring() got an unexpected keyword argument 'xml_declaration'

Looking into it, but have you seen it before?

**Reply by guangyee** [reply]
> Thank you for your interest in CI! That one is multi-layered:
> 
>     1. PR gating - currently, pull requests are gated by building the content for certain products on Jenkins nodes. Currently we have Fedora and RHEL nodes: https://jenkins.complianceascode.io/computer Those nodes now run test for SLE12 and SLE15 products, and those tests make sure that content built on those nodes passes sanity checks - i.e. that the build completes, and the result passes the validation.

Thanks for the information!

>        In the case that you would be so kind to host a VPS with a SLE system set up as a Jenkins node, we would gladly integrate it in our CI pipeline.

The decision to host a VPS node is beyond my pay grade at this point. But I'll convey the sentiment higher up.

> 
>     2. Rule tests - as you have probably noticed, rules can be tested individually - that's what the `tests` folder is there for. Rule tests are somehow slow, and there are many rules, so we don't do that as part of the PR gating. We have a bot that [advises users](https://github.com/ComplianceAsCode/content/pull/6453#issuecomment-738127318) to locally tests a PR using the upstream test suite - the bot created by our great engineer @mildas tries to estimate which tests are relevant in the context of a PR, and we hope that we will be able to run them as part of gating later.

Nice! Is the recommendation algorithm in the same repo?

>        The test suite may be somehow difficult to approach, and you can look forward a post going into greater detail on our [upstream blog](https://complianceascode.github.io/) in a couple of days. In a nutshell, the test suite is able to use OS container images or libvirt domains to provision the system using Bash scripts, and to execute the scan, collect the results, and propagate them to the user. So the idea is that the developer sets up the base system once, and then runs the test suite locally as a development support, or runs it overnight for all rules to makes sure that the datastream is in a good shape.

What was the reason to standardized on using Bash instead of Ansible for the test setup? Look forward to reading the blog.

**Reply by guangyee** [reply]
> @ggbecker trying to build the stig_overlay and hitting an error:
> vagrant@compliance-as-code:~/ComplianceAsCodeContent> PYTHONPATH=`./.pyenv.sh` ./utils/create-stig-overlay.py --disa-xccdf=/tmp/U_SLES_12_STIG_V2R1_Manual-xccdf.xml --ssg-xccdf=build/ssg-sle12-xccdf.xml -o sle12/overlays/stig_overlay.xml
> Traceback (most recent call last):
> File "./utils/create-stig-overlay.py", line 171, in
> main()
> File "./utils/create-stig-overlay.py", line 167, in main
> new_stig_overlay(disa_xccdftree, ssg_xccdftree, args.output_file)
> File "./utils/create-stig-overlay.py", line 121, in new_stig_overlay
> dom = xml.dom.minidom.parseString(ET.tostring(new_stig_overlay, encoding="UTF-8", xml_declaration=True))
> TypeError: tostring() got an unexpected keyword argument 'xml_declaration'
> 
> Looking into it, but have you seen it before?

Looks like `xml_declaration` keyword argument is new in 3.8 according to documentation.

https://docs.python.org/3/library/xml.etree.elementtree.html

Which means this exception will occur for versions prior to 3.8. Not sure if the Python version requirement is documented somewhere. If we intend to support older Python3 versions, we will need to catch that exception and try without the `xml_declaration` argument.

**Reply by brett060102** [reply]
@guangyee @ggbecker Thank you both. I can now build the overlays for both sle-12 and sle-15 on my system with python 3.8.5 using:
sampsone@bigubuntu:~/github/STIGS/merge/content$ PYTHONPATH=$(./.pyenv.sh) ./utils/create-stig-overlay.py --disa-xccdf=/tmp/U_SLES_15_STIG_V1R1_Manual-xccdf.xml --ssg-xccdf=build/ssg-sle15-xccdf.xml -o sle15/overlays/stig_overlay.xml

Generated the new STIG overlay file: sle15/overlays/stig_overlay.xml
sampsone@bigubuntu:~/github/STIGS/merge/content$ PYTHONPATH=$(./.pyenv.sh) ./utils/create-stig-overlay.py --disa-xccdf=/tmp/U_SLES_12_STIG_V2R1_Manual-xccdf.xml --ssg-xccdf=build/ssg-sle12-xccdf.xml -o sle12/overlays/stig_overlay.xml

Generated the new STIG overlay file: sle12/overlays/stig_overlay.xml

Did need to change PYTHONPATH=./.pyenv.sh  to PYTHONPATH=$(./.pyenv.sh) 

From what was in https://complianceascode.readthedocs.io/en/latest/manual/developer/04_updating_reference_and_overlay.html?highlight=create-stig-overlay#stig-overlay-content
Thank you both.

**Reply by ggbecker** [reply]
@brett060102 Please take a look at: https://github.com/ComplianceAsCode/content/pull/6666#issuecomment-795354388

I took the liberty and pushed some commits to the pull request. They add the initial stig_overlay.xml files and fix an issue in SLE15.

That's great you were able to make it work.

**Reply by brett060102** [reply]
@ggbecker thank you for all your work on this. I have https://github.com/SUSE/ComplianceAsCodeContent/pull/251 up which is a bulk enable of existing SLE-12 rules to tag them as SLE-15. No "new" content was created. Please tell me if we are on the right track PR wise.

**Reply by matejak** [reply]
> @guangyee wrote:
>
> The decision to host a VPS node is beyond my pay grade at this point. But I'll convey the sentiment higher up.

To put this in a context, our expenses for a VPS is about 150 € / year (including tax), so please do mention this higher up as well.

> Nice! Is the recommendation algorithm in the same repo?

Not yet, but you can check it out at https://github.com/mildas/content-test-filtering

> What was the reason to standardized on using Bash instead of Ansible for the test setup? Look forward to reading the blog.

That's an interesting question - when I think of it, I would even say that Ansible could do the same job with objectively simpler setup code.
Regarding the answer, it's probably an MVP that was good enough, so it stayed that way. The test env setup is [handled in one method](https://github.com/ComplianceAsCode/content/blob/master/tests/ssg_test_suite/rule.py#L395), so expanding possible formats is certainly doable.
The blog post is proposed as a PR https://github.com/ComplianceAsCode/ComplianceAsCode.github.io/pull/19 so you can check it out and provide any feedback that comes to your mind.

**Reply by ggbecker** [reply]
@guangyee @brett060102 @abergmann 

FYI, the blogpost about testing that @matejak has mentioned is now live and you can read it here: https://complianceascode.github.io/template/2021/03/25/tests_howto.html
